Proposal- Porch Performance Testing Framework - Operation Flow with Prometheus

mansoor17syed opened the pull request:
**Proposal: Porch Performance Testing Framework with Prometheus Integration**

**Overview**
This proposal introduces a comprehensive performance testing framework for Porch, designed to measure and monitor repository and package management operations using Prometheus. The framework provides automated setup, detailed metrics collection, and extensive reporting capabilities.

Basic Test Execution
```
go test -v ./... \
  --repos=2 \
  --packages=5 \
  --packageRevisions=4
  
```
**Goals**

1. Establish automated performance testing for Porch operations
2. Implement comprehensive metrics collection using Prometheus
3. Provide detailed operation monitoring and reporting
4. Enable configurable test scenarios

**Implementation Details**

1. Framework Components

- Automated Setup
- Prometheus Configuration
- Automatic container creation and initialization
- Custom prometheus.yml configuration
- Access: http://localhost:9090
- Metrics Server
- Port: :2113
- Endpoint: http://localhost:2113/metrics
- Logging System
- Directory: ./logs
- Format: porch-metrics-YYYY-MM-DD-HH-MM-SS.log
- Console and file logging

**2. Monitored Operations**

**Repository Management**

- Create Gitea Repository
- Create Porch Repository
- Wait Repository Ready
- Delete Repository

**Package Management**

- Create PackageRevision
- Update to Proposed
- Update to Published
- Delete PackageRevision

**Metrics Collection**

- Performance Metrics
- Operation duration tracking
- Success/failure counting
- Resource utilization monitoring
- Error tracking and reporting

**Prometheus Integration**

```
scrape_configs:
  - job_name: 'porch_metrics'
    static_configs:
      - targets: ['host.docker.internal:2113']
    scrape_interval: 1s
```

**Benefits**

- Automated performance testing
- Comprehensive metrics collection
- Real-time monitoring capabilities
- Detailed reporting and analysis
- Configurable test scenarios

**Future Enhancements**

- Additional metric types
- Enhanced visualization options
- Automated performance regression detection
- Extended test scenario support


![image](https://github.com/user-attachments/assets/70c36811-f024-487f-8f90-e5b87a0a0285)
![image](https://github.com/user-attachments/assets/a4bbc370-f5da-49b4-bc78-1afda72356e1)
![image](https://github.com/user-attachments/assets/3e0b571f-9250-4f56-b5ae-f5d4e0b07911)
![image](https://github.com/user-attachments/assets/db14cfa1-5455-40f5-ac16-a72fd60f3ec2)
![image](https://github.com/user-attachments/assets/a281de9d-49ec-4bb0-a7fa-80ffb7c4424a)
![image](https://github.com/user-attachments/assets/92a036d5-3c4d-46f4-ae80-3b5edbedb78f)
![image](https://github.com/user-attachments/assets/97f8e4e8-d9fd-42f5-bc45-cf6b33e48be3)
![image](https://github.com/user-attachments/assets/b16831ca-534a-4ac5-ae86-5736dada109c)

mansoor17syed commented:
Hi team,
I've worked on the Porch Performance Testing Framework with Prometheus integration. It includes automated setup, metrics collection, and detailed logging. Please review the changes and share your feedback.

efiacor commented:
Hi @mansoor17syed .
Thanks for setting this up. Really nice to have.
Couple of points. 
Is the plan for this to be run as an e2e test? If so, we should prob align with how the others are doing it.
See [here ](https://github.com/nephio-project/porch/blob/main/.github/workflows/porch-e2e.yaml)for the main e2e gh action workflow.
So, it sets up the kind cluster, deploys porch with the Make target "run-in-kind", and then executes the tests.
They also have a flag to avoid these being executed as part of the unit test runs - "E2E=1 go test ..."

For interacting with gitea, we should avoid using the ip address.
https://github.com/nephio-project/porch/pull/174/files#diff-da6008a39400b7b8b7c9a7b68d4b0928faf93f44c3ca48ad1d29cb9cfb4ae40fR37

The existing e2e suite use a stubbed git deployment - https://github.com/nephio-project/porch/blob/main/test/e2e/suite.go#L556
Could we make use of this for your suite?

liamfallon commented:
/retest

mansoor17syed commented:
Hi @efiacor 

Apologies for the delayed response—I was on vacation.

The main goal was to add metrics, but I wasn't sure how the community would receive them. So, I initially included them as part of the tests and proposed the changes.

I’m happy to modify or integrate these metrics directly into the code based on your feedback and requirements. Since you have a better perspective on this, I’d appreciate your guidance.

Looking forward to your response.

liamfallon commented:
/retest

liamfallon commented:
Hi @mansoor17syed , the unit tests are failing on this one.

efiacor commented:
> Hi @efiacor
> 
> Apologies for the delayed response—I was on vacation.
> 
> The main goal was to add metrics, but I wasn't sure how the community would receive them. So, I initially included them as part of the tests and proposed the changes.
> 
> I’m happy to modify or integrate these metrics directly into the code based on your feedback and requirements. Since you have a better perspective on this, I’d appreciate your guidance.
> 
> Looking forward to your response.

Hi @mansoor17syed ,

No problem. So if we want to add this as a stand alone suite we will need some mech to setup the test cluster.
The e2e suites use a local kind cluster and a stubbed gitea. If you need to use a full gitea deployment then we have a pkg for that.
Also, to avoid these being run during the unit tests, we will need to add some mech to skip similar to here - https://github.com/nephio-project/porch/blob/main/test/e2e/e2e_test.go#L62

mansoor17syed commented:
1. Regarding cluster setup:

For running these performance tests, can I use the existing kind cluster setup from the e2e framework? 
I notice the e2e tests use a local kind cluster - should I integrate with that same mechanism?

2. Regarding Gitea:

The e2e suites use a stubbed gitea deployment. You mentioned there's a pkg for full gitea deployment - which approach would be more appropriate for these performance tests? Should I:
a) Use the existing stubbed gitea from e2e suite
b) Use the full gitea deployment pkg you mentioned

3. Regarding test organization:

Would you prefer these performance tests to:
a) Run as part of the existing e2e test suite
b) Remain as a standalone suite but use the e2e infrastructure
c) Have a completely separate setup

Looking forward to your thoughts!

efiacor commented:
Hi @mansoor17syed . Would be good to pick this up again for R5 if you are willing. Thanks

mansoor17syed commented:
/retest

mansoor17syed commented:
/retest

mansoor17syed commented:
Hi team,

There were some test failures, which I have now fixed. I have also added the README.md file too.

Please review and share your feedback.

kispaljr commented:
/retest

efiacor commented:
Hi @mansoor17syed 
Could you do a rebase and maybe squash the commits to avoid the repetition of "Proposal- Porch Performance Testing Framework - Operation Flow with Prometheus" message. Thanks

efiacor commented on test/performance/porch_metrics_test.go:
```diff
@@ -0,0 +1,518 @@
+	// Propose the package
+	start = time.Now()
+	var pkg porchapi.PackageRevision
+	err = c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: newPkg.Name}, &pkg)
```
Should we include the gets in the time calc?
Same for other gets below.

efiacor replied:
But technically the "Get" is not part of the package proposal right?

mansoor17syed replied:
You are pointing out that we're including the Get operation in our timing calculation for the "Update to Proposed" metric. This is technically incorrect because:
The Get operation is just retrieving the current state of the package
The actual "proposal" operation is the Update that changes the lifecycle state
Including the Get in the timing makes the metric less accurate for measuring the actual proposal operation

The correct approach would be to:
First get the package (outside the timing)
Then start timing only the actual update operation

efiacor replied:
Apologies. My first comment was unclear. I was pointing out that we prob shouldn't include the Get operation in the timings, like you mention above.

mansoor17syed replied:
No worries will fix it

efiacor commented on test/performance/porch_metrics_test.go:
```diff
@@ -0,0 +1,518 @@
+	return metrics
+}
+
+func setupMonitoring(t *testing.T) error {
```
Would it not be better to deploy a basic in memory k8s prometheus pod, etc?
We should have these configs somewhere we can tweak them if needed.

mansoor17syed replied:
Yes, you're right—this is a good solution. and this is exactly what I have in mind.

To implement this change, we would need to:

Create a Prometheus ConfigMap

Create a Prometheus Deployment

Create a Service to expose Prometheus

Use the Kubernetes client to manage these resources

 However, the goal of this PR is to initially implement metrics and capture them as part of test or performance test. If this approach works well, the next step would be to integrate it into the code so that each operation generates metrics. These metrics would then be collected using an external Prometheus instance. This aligns with what the service assurance team is aiming to achieve.

efiacor commented on test/performance/porch_metrics_test.go:
```diff
@@ -0,0 +1,518 @@
+		cmd  string
+		args []string
+	}{
+		{"network create", "docker", []string{"network", "create", "monitoring"}},
```
Not a fan of this direct docker calls.

efiacor commented on test/performance/porch_metrics_test.go:
```diff
@@ -0,0 +1,518 @@
+		Spec: configapi.RepositorySpec{
+			Type: "git",
+			Git: &configapi.GitRepository{
+				Repo:   fmt.Sprintf("http://172.18.255.200:3000/nephio/%s", repoName),
```
I think the URL should be configurable

efiacor commented on test/performance/porch_metrics_test.go:
```diff
@@ -0,0 +1,518 @@
+	start := time.Now()
+
+	// Create Gitea repo
+	err := createGiteaRepo(repoName)
```
Do we care how long it takes gitea to create the repo?

efiacor commented on test/performance/metrics.go:
```diff
@@ -0,0 +1,162 @@
+
+	return nil
+}
+func waitForRepository(ctx context.Context, c client.Client, t *testing.T, namespace, name string, timeout time.Duration) error {
```
*waitForPorchRepository

efiacor commented on test/performance/metrics.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+func createGiteaRepo(repoName string) error {
+	giteaURL := "http://172.18.255.200:3000/api/v1/user/repos"
```
Again, URL should be dynamic. Alternatively, we use a gitea client - 
https://pkg.go.dev/code.gitea.io/sdk/gitea#Client.CreateRepo

efiacor commented on test/performance/README.md:
````diff
@@ -0,0 +1,80 @@
+    --from-literal=password=secret
+```
+
+## 5. Define Gitea Repositories in Porch
````
Actually, I don't think this is necessary

efiacor commented on test/performance/README.md:
````diff
@@ -0,0 +1,80 @@
+E2E=1 go test -v ./... -repos=4 -packages=4 -timeout 60m
+```
+
+Parameters:
````
Indicate all possible inputs and default values.

mansoor17syed replied:
There are no default params and these are the only two possible params for now to run tests

efiacor replied:
I think there are default values - https://github.com/nephio-project/porch/pull/174/files#diff-2108193b521a1a0005d0859c990a880d78a8b44b726992a80c7b69784842b11eR38

efiacor commented on test/performance/README.md:
```diff
@@ -0,0 +1,80 @@
+- `repos`: Number of repositories to test
+- `packages`: Number of package revisions in each repository
+
+Once the tests begin, the metrics can be observed.
```
Give an example of the output and details of the log files created.

mansoor17syed replied:
Done

efiacor commented on test/performance/README.md:
````diff
@@ -0,0 +1,80 @@
+Retrieve the captured metrics using:
+```bash
+curl -k http://localhost:2113/metrics
+```
````
Add info regarding the queries file and how to use them

mansoor17syed replied:
yes there is a query file and data/info about promql attached to the git PR

efiacor commented on test/performance/porch_metrics_test.go:
```diff
@@ -0,0 +1,518 @@
+			Git: &configapi.GitRepository{
+				Repo:   fmt.Sprintf("http://172.18.255.200:3000/nephio/%s", repoName),
+				Branch: "main",
+				SecretRef: configapi.SecretRef{
```
We could define this secret locally and create it as part of the test setup.

efiacor replied:
Not sure what your reply refers to. We can define a secret in a yaml file in the same dir and creat it as part of the test setup.

efiacor commented on test/performance/porch_metrics_test.go:
```diff
@@ -0,0 +1,518 @@
+		start = time.Now()
+		err = waitForRepository(ctx, c, t, namespace, repoName, 60*time.Second)
+		duration = time.Since(start).Seconds()
+		recordMetric("Wait Repository Ready", repoName, "", duration, err)
```
... Porch Repository Ready ...

mansoor17syed replied:
Done

mansoor17syed commented on test/performance/README.md:
````diff
@@ -0,0 +1,80 @@
+## 5. Define Gitea Repositories in Porch
+Apply the repository configuration file:
+```bash
+kubectl apply -f porch-repositories.yaml
````
Corrected

mansoor17syed commented on test/performance/README.md:
````diff
@@ -0,0 +1,80 @@
+## 7. Run Performance Tests
+Navigate to the performance test directory:
+```bash
+cd porch/test/performance/
````
Corrected

efiacor commented on test/performance/porch_metrics_test.go:
```diff
@@ -0,0 +1,532 @@
+
+	// Delete package
+	start = time.Now()
+	err = c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: pkg.Name}, &pkg)
```
All of the "Get"s should be outside the timers.

mansoor17syed replied:
Done

efiacor commented:
/approve
/lgtm
